DMD-285 Table profiling: New numeric stats metric

martinsifra opened the pull request:
**Jira**: DMD-285

**Connection:** keboola/connection#6099
**SAPI**: keboola/storage-api-php-client#1541
**Snowflake drive:** keboola/storage-backend#223

martinsifra committed:
test(table profiling) Add missing "profile" field assert

- DMD-102
- Add missing $response->getProfile() assert.

martinsifra committed:
feat(table profiling) New metric with numeric statistics

- DMD-285
- New "NumericStatisticsColumnMetric" returning average, mode, median,
  min and max values.
- Metric has commented tests for untyped tables because statistics
  functions require only numeric values. How to cast strings in
  untyped tables to numeric will be solved in the future.
- Warning: Median is only single value even there are multiple
  ones with the same frequency.

martinsifra committed:
feat(table profiling) Metrics by column type

- DMD-285
- Column metrics are selected by column type.
  - STRING column has extra AvgMinMaxLengthColumnMetric.
  - INTEGER column has extra NumericStatisticsColumnMetric.

martinsifra commented on src/Handler/Table/Profile/ProfileTableHandler.php:
```diff
@@ -89,35 +92,57 @@ public function __invoke(
+
+        $extra = match ($type) {
+            Bigquery::TYPE_NUMERIC,
+            'FLOAT',
```
Z BQ padá `FLOAT` ale my máme u nás jen `FLOAT64` 😟 Asi by to chtělo přidat do té libky v monorepu, ale nějak postupně. Snažil jsem se to hledat i v dokumentaci, ale možný je používat oboje a nevím proč, když pošlu při vytváření tabulky `FLOAT64` dostávám tady jen `FLOAT`.

jirkasemmler replied:
nenapadlo te pouzit basetype? vyrobit instanci `Keboola\Datatype\Definition\Bigquery` a zavolat si `getBasetype`

to same plati pak i pro SNFLK

martinsifra replied:
Vyřešeno BaseTypem a doplněním FLOAT a BOOLEAN aliasů do `php-basetypes` https://github.com/keboola/storage-backend/pull/231.

martinsifra commented on src/Profile/Column/NumericStatisticsColumnMetric.php:
```diff
@@ -0,0 +1,104 @@
+         *     }
+         * } $results
+         */
+        $results = iterator_to_array($context->client->runQuery($context->client->query($sql)));
```
Tady ve SNFLK by to možná chtělo nějaký try-catch a re-throw. Když neprojde ta query, mohlo by se to samo zotavit a nic to nevracet. Zároveň by to ale mělo logovat, abychom to někde viděli, že to padá.

jirkasemmler replied:
tak to do retry

martinsifra replied:
Je tam try-catch s logováním, to by mělo asi stačit.

martinsifra committed:
feat(table profiling) Try-catch when executing metrics

- DMD-357
- Wrap all calls to BigQuery API to prevent any profiling failure
  cause by wrong API call.

martinsifra commented on src/Handler/Table/Profile/ProfileTableHandler.php:
```diff
@@ -81,43 +87,88 @@ public function __invoke(
+                $this->internalLogger->warning(
+                    $e->getMessage(),
+                    [
+                        'exception' => $e->getPrevious(),
```
Jde předat celá exception, JsonFormatter v Monologu má  sobě encodování Exceptions. Kluci z PS to takhle používají.
